src/MatrixProductOperator.jl

```diff
@@ -0,0 +1,168 @@
+#using OptimizedEinsum: ContractionPath
+using UUIDs: uuid4
+using IterTools: partition
+using Random
+
+abstract type MatrixProductOperator{B} <: State{B} end
+
+function MatrixProductOperator(arrays; bounds::Type{<:Bounds} = Open, kwargs...)
+    MatrixProductOperator{bounds}(arrays; kwargs...)
+end
+
+function MatrixProductOperator{Open}(arrays; χ = nothing, order = (:l, :r, :i, :o), meta...)
+    !issetequal(order, (:l, :r, :i, :o)) && throw(ArgumentError("`order` must be a permutation of the :l, :r, :i and :o"))
+    order = Dict(dim => i for (i, dim) in enumerate(order))
+
+    # check format
+    # NOTE matching index dimensions are checked on `push!(::TensorNetwork,...)`
+    # TODO check χ?
+    if !all(==(4) ∘ ndims, arrays[2:end-1]) || ndims(first(arrays)) != 3 || ndims(last(arrays)) != 3
+        throw(DimensionMismatch("virtual bond-dims mismatch"))
+    end
+
+    tn = TensorNetwork{MatrixProductOperator{Open}}(; χ = χ, order = order, meta...)
+
+    n = length(arrays)
+    vinds = Dict(x => Symbol(uuid4()) for x in partition(1:n, 2, 1))
+    oinds = Dict(i => Symbol(uuid4()) for i in 1:n)
+    iinds = Dict(i => Symbol(uuid4()) for i in 1:n)
+    permutator = [order[i] for i in (:l, :r, :i, :o)]
+
+    # add boundary tensors
+    # first
+    labels = permute!([vinds[(1,2)], iinds[1], oinds[1]], [order[:r], order[:i], order[:o]])
+    push!(tn, Tensor(first(arrays), labels))
+
+    # last
+    labels = permute!([vinds[(n-1,n)], iinds[n], oinds[n]], [order[:l], order[:i], order[:o]])
```

Comment:
The previous code raised an error since you are doing `permute!` over a size-3 vector with permutations from 1 to 4. I suggest:
```suggestion
    labels = permute!([vinds[(1,2)], vinds[(1,2)], iinds[1], oinds[1]], [order[:l], order[:r], order[:i], order[:o]])[2:end]
    push!(tn, Tensor(first(arrays), labels))

    # last
    labels = permute!([vinds[(n-1,n)], vinds[(n-1,n)], iinds[n], oinds[n]], [order[:r], order[:l], order[:i], order[:o]])[2:end]
```

Reply:
Solving it at the moment. Let me check.

Reply:
Although functional, here I'm leaving a more (what I think) understandable approach.
```suggestion
    permutator = [order[:r], order[:i], order[:o]]
    permutator .= permutator .- minimum(permutator) .+ 1
    k = only(setdiff(1:4, permutator))
    permutator[permutator .> k] .-= 1

    labels = permute!([vinds[(1,2)], iinds[1], oinds[1]],  permutator)
    push!(tn, Tensor(first(arrays), labels))

    # last
    permutator = [order[:l], order[:i], order[:o]]
    permutator .= permutator .- minimum(permutator) .+ 1
    k = only(setdiff(1:4, permutator))
    permutator[permutator .> k] .-= 1

    labels = permute!([vinds[(n-1,n)], iinds[n], oinds[n]], permutator)
```

Reply:
Going to add the function in `Helper.jl` to avoid repeating the code